vars/artifactSetVersion.groovy

```diff
@@ -30,7 +30,7 @@ def call(Map parameters = [:], Closure body = null) {
     handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: parameters) {
 
         def gitUtils = parameters?.juStabGitUtils ?: new GitUtils()
-        def script = parameters?.script ?: [commonPipelineEnvironment: commonPipelineEnvironment]
+        def script = parameters?.script ?: this
```

Comment:
Can anybody describe a valid use case where it is not possible to hand over the reference to the pipeline script itself to a step (```<stepname> script: this```). If not: not having ```parameters.script``` is a mistake. What is the benefit of implementing a fallback. Wouln't it be better to raise an exception? That would give the chance to the developer to fix the mistake rather than continuing with a fallback where nobody is aware of. Basically such kind of a mistake (forgetting about ```script: this``` when calling a step) would be detected soon during development provided we check for it and raise an exception.

Other way: when ```this``` is a valid approach: Why using the pipeline script (```<stepname> script: this``` / ```parameters.script```) at all? If using the step itself does the same job like using the pipeline script why should we use the pipeline script? In this case we can simply say ```def script = this``` always (or maybe even inline).

My understanding up to now is: the pipeline script should be used, not the step(s). With that each step shares the same environment

--> Would be good if somebody could explain the difference between using the pipeline step and using the pipeline script (aka: ```Jenkinsfile```).

Reply:
I think, project piper had the approach that you can use just single steps, without calling setupPipelineEnvironment before. 
Only if you call setupPipelineEnvironment, you have to pass the script, because it contains the configuration and other values which were initialized in the setupPipelineEnvironment.

`script=this` results in an empty configuration, as `script=[commonPipelineEnvironment: commonPipelineEnvironment]`

Reply:
We have two use cases for the script:
1. access general steps (archive, stash, echo, error)
2. access the global configuration (which is also a step)

For the first it is not importalt if we use `this` from a step or the Jenkinsfile.

Reply:
ad 1.) Here it does not matter. Both ways (using the pipeline script or using the step) provides access to e.g. ```echo```, ```sh```...

ad 2.) Here is makes a difference. Each step comes with another instance of commonPipelineEnvironment. And beside this we have the commonPipelineEnvironment instance bound the the pipeline script itself. The commonPipelineEnvironment bound to the script itself is the one with is normally propery configured by a prior setupCommonPipelineEnvironment call (if any). Sharing configuration values is not possible when we use the various commonPipelineEnvironment instances bound to all the steps. Hence using ```this``` like ```def script = this``` does not work wrt configuration.

Wouldn't it be better not to have ```def script = this``` since it fails with 2.)? IMO having a mandatory ```script``` in the parameters map providing a reference to the pipeline script itself and in turn providing a reference to the shared commonPipelineEnvironment is clearly better.

@daniel-kurzynski 
>you can use just single steps, without calling setupPipelineEnvironment before

There is IMO nothing wrong with having ```script``` in the signature holding a reference to the surrounding pipeline script, even if ```setupCommonPipelineEnvironment``` was not called.